docs/proposals/03-networking.md

```diff
@@ -0,0 +1,72 @@
+# Network Extensibility
+
+Currently Gardener follows a mono network-plugin support model (i.e., Calico). Although this can seem to be the more stable approach, it does not completely reflect the real use-case. This proposal brings forth an effort to add an extra level of customizability to Gardener networking.
+
+## Motivation
+
+Gardener is an open-source project that provides a nested user model. Basically, there are two types of services provided by Gardener to its users:
+
+- **Managed**: users only request a Kubernetes cluster (Clusters-as-a-Service)
+- **Hosted**: users utilize Gardener to provide their own managed version of Kubernetes (Cluster-Provisioner-as-a-service)
+
+For the first set of users, the choice of network plugin might not be so important, however, for the second class of users (i.e., Hosted) it is important to be able to customize networking based on their needs.
+
+Furthermore, Gardener provisions clusters on different cloud-providers with different networking requirements. For example, Azure does not support Calico Networking [1], this leads to the introduction of manual exceptions in static add-on charts which is error prune and can lead to failures during upgrades.
+
+Finally, every provider is different, and thus the network always needs to adapt to the infrastructure needs to provider better performance. Consistency does not necessarily lie in the implementation but in the interface.
+
+## Gardener Network Extension
+
+The goal of the Gardener Network Extensions is to support different network plugin, therefore, the specification for the network resource won't be fixed and will be customized based on the underlying network plugin. To do so, a `NetworkConfig` field in the spec will be provided where each plugin will define. Below is an example for deploy Calico as the cluster network plugin.
+
+```yaml
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Network
+metadata:
+  name: calico-network
+  namespace: shoot--core--test-01
+spec:
+  type: calico
+  clusterCIDR: 192.168.0.0/24
+  serviceCIDR:  10.96.0.0/24
+  providerConfig:
+    apiVersion: calico.extensions.gardener.cloud/v1alpha1
+    kind: NetworkConfig
+    ipam:
+      type: host-local
+      cidr: usePodCIDR
+    backend: bird
+status:
+  observedGeneration: ...
+  state: ...
+  lastError: ..
+  lastOperation: ...
+  providerStatus:
+    apiVersion: calico.extensions.gardener.cloud/v1alpha1
+    kind: NetworkStatus
```

Comment:
Who uses this information/what is it good for?

Reply:
Monitoring, this can be exported for metrics or used for network validation testing. Also, useful for feedback on how networking in a cluster is doing in general.